### PR TITLE
  Implementation of ConcaveHull using simple Chi-Hull algorithm.

### PR DESCRIPTION
**Pull request by grootstebozewolf**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/NetTopologySuite/NetTopologySuite/pulls) open
- [ ] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository.

- [x] I have provided test coverage for my change (where applicable)

### Description
Computes a 𝛘-shape of a Geometry which is a concave Geometry when the threshold is small enough and a convex geometry with a large threshold that contains all the points in the input Geometry. Uses the algorithm defined in [Efficient generation of simple polygons for characterizing the shape of a set of points in the plane](https://www.sciencedirect.com/science/article/abs/pii/S0031320308001180).

### CODE REVIEW

**Comment by FObermaier** [comment]
@grootstebozewolf , thanks for your contribution.
The `ConvexHull` stub in JTS/NTS has bothered me for quite some time, too. That's why I ported 
mourners `concaveman` algorithm (https://github.com/locationtech/jts/pull/427). A decision to accept is has not made so far.

From the discussion, I gather that we should not name it `ConcaveHull` but give it some other name describing the algorithm used. If that is met, I think we can merge.

**Comment by grootstebozewolf** [comment]
Maybe we need a strategy pattern? This algorithm is also known as a 𝛘-shape, or ChiShape.

**Comment by airbreather** [comment]
I just pushed a few changes that bring back the old versions of the existing files, since those don't need to be marked as modified for this anymore after the rename.

I'll do a review of this soon.  Sorry it's taken so long, most of my NTS time has been going into trying to tidy up v2 stuff post-release and respond to user feedback about it.

> Maybe we need a strategy pattern? This algorithm is also known as a 𝛘-shape, or ChiShape.

I think that's where the discussion started on the other issue that @FObermaier linked to earlier in this thread.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                                              where n != null && n.IsLive()
+                                              select n)
+                    {
+                        Debug.WriteLine(neighbour);
+                        CheckMaxEdgeAndListBorderTriangles(quadEdgeTriangleList, neighbour);
+                    }
+                }
+            }
+            var lineMerger = new LineMerger();
+            var fact = new GeometryFactory();
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: This should use `_geom.Factory` instead of making a new one.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+namespace NetTopologySuite.Hull
+{
+    public class ChiShape
+    {
+        private readonly Geometry _geom;
+        private readonly double _tolerance;
+
+        public ChiShape(Geometry geom, double tolerance)
+        {
+            _geom = geom;
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]:
```csharp
_geom = geom ?? throw new ArgumentNullException(nameof(geom));
```

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                CheckMaxEdgeAndListBorderTriangles(quadEdgeTriangleList, triangle);
+            }
+
+            while (quadEdgeTriangleList.Count != 0)
+            {
+                var triangle = quadEdgeTriangleList.Last();
+                var edge = triangle.Key;
+                Debug.WriteLine(triangle.Value);
+                quadEdgeTriangleList.Remove(edge);
+                if (edge.Length.CompareTo(_tolerance) >= 0 && RemovalGivesRegularPolygon(triangle.Value))
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: :question: shouldn't we be breaking out of the loop once we find an edge such that `edge.Length < _tolerance`?  Once we see the first edge that fails the first condition, then I think all other edges will fail too...

Something like:
```csharp
if (edge.Length < _tolerance)
{
    break;
}

if (!RemovalGivesRegularPolygon(triangle.Value))
{
    continue;
}

var neighbours = triangle.Value.GetNeighbours();
// etc...
```

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                }
+            }
+            var lineMerger = new LineMerger();
+            var fact = new GeometryFactory();
+            foreach (var triangle in tris)
+            {
+                if (!triangle.IsLive()) continue;
+                for (int i = 0; i < 3; i++)
+                {
+                    if (triangle.GetAdjacentTriangleAcrossEdge(i) == null || !triangle.GetAdjacentTriangleAcrossEdge(i).IsLive())
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: nit: repeated call to `GetAdjacentTriangleAcrossEdge` can be eliminated:
```csharp
if (triangle.GetAdjacentTriangleAcrossEdge(i)?.IsLive() != true)
```

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            var subDiv = builder.GetSubdivision();
+            return subDiv;
+        }
+    }
+
+    internal class QuadEdgeComparer : Comparer<QuadEdge>
+    {
+        public override int Compare(QuadEdge x, QuadEdge y)
+        {
+            return x.Length.CompareTo(y.Length);
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: The `IComparer<TKey>` used for a `SortedList<TKey, TValue>` is supposed to give a total ordering.  This only checks the `Length`, which means that if two different `QuadEdge`s happen to have the same length, then only one of them will be allowed to live in the list at a given time.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+{
+    public class ChiShape
+    {
+        private readonly Geometry _geom;
+        private readonly double _tolerance;
+
+        public ChiShape(Geometry geom, double tolerance)
+        {
+            _geom = geom;
+            _tolerance = tolerance;
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Probably also validate that `tolerance` is a finite, positive number.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        private static IList<QuadEdgeTriangle> ExtractTriangles(QuadEdgeSubdivision subdiv)
+        {
+            var qeTris = QuadEdgeTriangle.CreateOn(subdiv);
+            return qeTris;
+        }
+
+
+        private Geometry ComputeHull(IList<QuadEdgeTriangle> tris)
+        {
+            var quadEdgeTriangleList = new SortedList<QuadEdge, QuadEdgeTriangle>(new QuadEdgeComparer());
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: After going through everything else, I'm not sure `SortedList<TKey, TValue>` is the right thing to use here.  I'm favoring an `AlternativePriorityQueue<TPriority, TData>` with a `Dictionary<TKey, TValue>` on the side.

- The way we would have to contort our usage of `SortedList<TKey, TValue>` to eliminate an `O(n²)` bit is unnatural.  We only extract values on one end of the sorted list, which is what priority queues optimize for.
- The criteria we should be using for `.ContainsKey` is different from the criteria we use for sorting, which is incompatible with `SortedList<TKey, TValue>`.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            foreach (var triangle in tris)
+            {
+                CheckMaxEdgeAndListBorderTriangles(quadEdgeTriangleList, triangle);
+            }
+
+            while (quadEdgeTriangleList.Count != 0)
+            {
+                var triangle = quadEdgeTriangleList.Last();
```
[the remark this answers is not in the document]
Comment (airbreather) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: The best I can come up with, looking at just this part of the code, is this `O(log(n))` bit, which I think looks awful:
```csharp
var edge = quadEdgeTriangleList.Keys.Last();
var triangle = quadEdgeTriangleList[edge];
```

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                if (!triangle.IsLive()) continue;
+                for (int i = 0; i < 3; i++)
+                {
+                    if (triangle.GetAdjacentTriangleAcrossEdge(i) == null || !triangle.GetAdjacentTriangleAcrossEdge(i).IsLive())
+                    {
+                        lineMerger.Add(triangle.GetEdge(i).ToLineSegment().ToGeometry(fact));
+                    }
+                }
+            }
+            var mergedLine = (LineString)lineMerger.GetMergedLineStrings().FirstOrDefault();
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: nit: these two lines look like they're arguing about whether or not the list is actually allowed to be empty
- if the list **can** be empty, then `if (mergedLine.IsRing)` will throw `NullReferenceException`
- if the list **can't** be empty, then `.FirstOrDefault()` should be more like `[0]`

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            if (mergedLine.IsRing)
+            {
+                var lr = new LinearRing(mergedLine.CoordinateSequence, fact);
+
+                var concaveHull = new Polygon(lr, null, fact);
+
+                return concaveHull;
+            }
+            return mergedLine;
```
Comment (airbreather) [inline src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Documentation says that this is supposed to throw if we can't generate a hull, but this returns a 1-dimensional geometry... which is the correct behavior?

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 9 above)
Comment (airbreather) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: follow-up: "these two lines" was supposed to be a multi-line comment targeting 86 and 87... apparently multi-line comments are still in beta

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Not sure if the SortedList makes this O(n²), we should test this in a lab environment :)

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 4 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: True

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 5 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Do we need to solve this edge case :)

Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Can you help me with creating a test for this, not sure when this would be an issue.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 6 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Check!

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 7 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: We will leave this as an exercise for the viewers.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 9 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: I guess this is another edge case, when the geometry only has one or two points we should not throw something but return a defualt null value.

File: src/NetTopologySuite.Lab/Hull/ChiShape.cs (hunk 10 above)
Comment (grootstebozewolf) [reply src/NetTopologySuite.Lab/Hull/ChiShape.cs]: Let's just throw an exception and call it a day.
